File: app/src/main/kotlin/com/albertferran/eatapp/ui/common/CuisineVisuals.kt
package com.albertferran.eatapp.ui.common

import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.BakeryDining
import androidx.compose.material.icons.filled.Bento
import androidx.compose.material.icons.filled.BreakfastDining
import androidx.compose.material.icons.filled.BrunchDining
import androidx.compose.material.icons.filled.DinnerDining
import androidx.compose.material.icons.filled.Fastfood
import androidx.compose.material.icons.filled.Flatware
import androidx.compose.material.icons.filled.Grass
import androidx.compose.material.icons.filled.Icecream
import androidx.compose.material.icons.filled.KebabDining
import androidx.compose.material.icons.filled.LocalBar
import androidx.compose.material.icons.filled.LocalCafe
import androidx.compose.material.icons.filled.LocalDining
import androidx.compose.material.icons.filled.LocalPizza
import androidx.compose.material.icons.filled.LunchDining
import androidx.compose.material.icons.filled.OutdoorGrill
import androidx.compose.material.icons.filled.RamenDining
import androidx.compose.material.icons.filled.Restaurant
import androidx.compose.material.icons.filled.RiceBowl
import androidx.compose.material.icons.filled.RoomService
import androidx.compose.material.icons.filled.SetMeal
import androidx.compose.material.icons.filled.SoupKitchen
import androidx.compose.material.icons.filled.SportsBar
import androidx.compose.material.icons.filled.Tapas
import androidx.compose.material.icons.filled.WineBar
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.res.stringResource
import com.albertferran.eatapp.data.local.Cuisine
import com.albertferran.eatapp.ui.theme.CuisineTint
import com.albertferran.eatapp.ui.theme.LocalCuisineAccents

/**
 * Since restaurants have no real photo data, list/detail screens use a cuisine-derived
 * icon and tint in its place instead of an image.
 *
 * All three helpers take the raw `cuisineType` string straight off the entity and
 * degrade gracefully when it isn't a key this build knows: generic icon, neutral
 * tint, and the raw value as the label.
 */
fun cuisineIcon(cuisineType: String): ImageVector = when (Cuisine.fromKey(cuisineType)) {
    Cuisine.MEDITERRANEAN -> Icons.Filled.LocalDining
    Cuisine.SPANISH -> Icons.Filled.Tapas
    Cuisine.CATALAN -> Icons.Filled.Flatware
    Cuisine.BASQUE -> Icons.Filled.RoomService
    Cuisine.ITALIAN -> Icons.Filled.LocalPizza
    Cuisine.JAPANESE -> Icons.Filled.RamenDining
    Cuisine.CHINESE -> Icons.Filled.RiceBowl
    Cuisine.ASIAN -> Icons.Filled.Bento
    Cuisine.INDIAN -> Icons.Filled.SoupKitchen
    Cuisine.MIDDLE_EASTERN -> Icons.Filled.KebabDining
    Cuisine.AMERICAN -> Icons.Filled.LunchDining
    Cuisine.SEAFOOD -> Icons.Filled.SetMeal
    Cuisine.BAR -> Icons.Filled.LocalBar
    Cuisine.BEER_BAR -> Icons.Filled.SportsBar
    Cuisine.WINE_BAR -> Icons.Filled.WineBar
    Cuisine.CAFE -> Icons.Filled.LocalCafe
    Cuisine.BAKERY -> Icons.Filled.BakeryDining
    Cuisine.DESSERT -> Icons.Filled.Icecream
    Cuisine.BREAKFAST -> Icons.Filled.BreakfastDining
    Cuisine.BRUNCH -> Icons.Filled.BrunchDining
    Cuisine.GRILL -> Icons.Filled.OutdoorGrill
    Cuisine.FAST_FOOD -> Icons.Filled.Fastfood
    Cuisine.FINE_DINING -> Icons.Filled.DinnerDining
    Cuisine.VEGETARIAN -> Icons.Filled.Grass
    null -> Icons.Filled.Restaurant
}

/** Localized display label, falling back to the raw stored value. */
@Composable
fun cuisineLabel(cuisineType: String): String =
    Cuisine.fromKey(cuisineType)?.let { stringResource(it.labelRes) } ?: cuisineType

@Composable
fun cuisineTint(cuisineType: String): CuisineTint {
    // Keyed off the enum ordinal rather than the string's hash: stable across
    // releases, and spread evenly over the palette's accents instead of landing
    // arbitrarily. An unknown key stays neutral so it reads as "no category"
    // rather than borrowing a colour that means something else.
    val ordinal = Cuisine.fromKey(cuisineType)?.ordinal
        ?: return CuisineTint(
            container = MaterialTheme.colorScheme.surfaceVariant,
            onContainer = MaterialTheme.colorScheme.onSurfaceVariant
        )
    return LocalCuisineAccents.current[ordinal]
}

File: lib/core/widgets/cuisine_visuals.dart
import 'package:flutter/material.dart';

import '../../data/models/cuisine.dart';
import '../l10n/generated/app_localizations.dart';
import '../theme/tokens/cuisine_accents.dart';

/// The icon, label and tint a cuisine badge is drawn with.
///
/// Restaurants carry no photo of their own yet, so the list and detail screens
/// draw a cuisine-derived badge in its place. All three lookups take the raw
/// `cuisineType` value straight off the entity and degrade gracefully when it
/// is not a key this build knows: a generic icon, a neutral tint and the raw
/// string as the label — so a newer data file never breaks an older app.
///
/// Ported from the Android app's `ui/common/CuisineVisuals.kt`.
IconData cuisineIcon(String cuisineType) =>
    switch (Cuisine.fromKey(cuisineType)) {
      Cuisine.mediterranean => Icons.restaurant,
      Cuisine.spanish => Icons.tapas,
      Cuisine.catalan => Icons.restaurant_menu,
      Cuisine.basque => Icons.room_service,
      Cuisine.italian => Icons.local_pizza,
      Cuisine.japanese => Icons.ramen_dining,
      Cuisine.chinese => Icons.rice_bowl,
      Cuisine.asian => Icons.takeout_dining,
      Cuisine.indian => Icons.soup_kitchen,
      Cuisine.middleEastern => Icons.kebab_dining,
      Cuisine.american => Icons.lunch_dining,
      Cuisine.seafood => Icons.set_meal,
      Cuisine.bar => Icons.local_bar,
      Cuisine.beerBar => Icons.sports_bar,
      Cuisine.wineBar => Icons.wine_bar,
      Cuisine.cafe => Icons.local_cafe,
      Cuisine.bakery => Icons.bakery_dining,
      Cuisine.dessert => Icons.icecream,
      Cuisine.breakfast => Icons.breakfast_dining,
      Cuisine.brunch => Icons.brunch_dining,
      Cuisine.grill => Icons.outdoor_grill,
      Cuisine.fastFood => Icons.fastfood,
      Cuisine.fineDining => Icons.dinner_dining,
      Cuisine.vegetarian => Icons.grass,
      null => Icons.restaurant,
    };

/// The translated label for a cuisine, or the raw stored value when the key is
/// unknown.
String cuisineLabel(AppLocalizations l10n, String cuisineType) {
  final Cuisine? cuisine = Cuisine.fromKey(cuisineType);
  return cuisine == null ? cuisineType : cuisine.label(l10n);
}

/// The accent pair for a cuisine badge.
///
/// Keyed off the enum's index rather than the string's hash: stable across
/// releases, and spread evenly over the palette's accents instead of landing
/// arbitrarily. An unknown key stays neutral so it reads as "no category"
/// rather than borrowing a colour that means something else.
CuisineTint cuisineTint(BuildContext context, String cuisineType) {
  final Cuisine? cuisine = Cuisine.fromKey(cuisineType);
  if (cuisine == null) {
    final ColorScheme scheme = Theme.of(context).colorScheme;
    return CuisineTint(
      container: scheme.surfaceContainerHighest,
      onContainer: scheme.onSurfaceVariant,
    );
  }
  return CuisineAccents.of(context)[cuisine.index];
}
